src/cascadia/TerminalApp/TerminalPage.cpp

```diff
@@ -994,6 +927,184 @@ namespace winrt::TerminalApp::implementation
         _newTabButton.Flyout(newTabFlyout);
     }
 
+    // Method Description:
+    // - For a given list of tabmenu entries, this method will create the corresponding
+    //   list of flyout items. This is a recursive method that calls itself when it comes
+    //   across a folder entry.
+    std::vector<WUX::Controls::MenuFlyoutItemBase> TerminalPage::_CreateNewTabFlyoutItems(IVector<NewTabMenuEntry> entries)
+    {
+        std::vector<WUX::Controls::MenuFlyoutItemBase> items;
+
+        if (entries == nullptr || entries.Size() == 0)
+        {
+            return items;
+        }
+
+        for (const auto& entry : entries)
+        {
+            if (entry == nullptr)
+            {
+                continue;
+            }
+
+            switch (entry.Type())
+            {
+            case NewTabMenuEntryType::Separator:
+            {
+                items.push_back(WUX::Controls::MenuFlyoutSeparator{});
+                break;
+            }
+            // A folder has a custom name and icon, and has a number of entries that require
+            // us to call this method recursively.
+            case NewTabMenuEntryType::Folder:
+            {
+                const auto folderEntry = entry.as<FolderEntry>();
+
+                auto folderItem = WUX::Controls::MenuFlyoutSubItem{};
+                folderItem.Text(folderEntry.Name());
+
+                auto icon = _CreateNewTabFlyoutIcon(folderEntry.Icon());
+                folderItem.Icon(icon);
+
+                // Recursively generate flyout items
+                auto folderEntries = _CreateNewTabFlyoutItems(folderEntry.Entries());
+                for (auto const& folderEntry : folderEntries)
+                {
+                    folderItem.Items().Append(folderEntry);
+                }
+
+                items.push_back(folderItem);
+                break;
+            }
+            // Any "collection entry" will simply make us add each profile in the collection
+            // separately. This collection is stored as a map <int, Profile>, so the correct
+            // profile index is already known.
+            case NewTabMenuEntryType::RemainingProfiles:
+            {
+                const auto remainingProfilesEntry = entry.as<ProfileCollectionEntry>();
+                if (remainingProfilesEntry.Profiles() == nullptr)
```

Comment:
Curious about the architectural division between having the settings model (ideally "dumb" objects) decide what is in this entry versus the application _consuming_ the settings model. It's an interesting conundrum. Your gut feeling is to do it in TSM -- why?

(I'm not saying it should be moved, but I am always curious about stuff like this!)

Reply:
I thought about this, and if I remember correctly (but you're really pushing my memory here :P ):
- To put it in the consuming side felt very dirty since that part was mostly UI rendering stuff and I wanted to provide complete entities to easily consume. In addition, I don't think most of the matching/profiles lookup is _as easy_ to do there, but the main reason is again that it just felt wrong.
- Then I believe I simply looked at how for example profiles themselves are resolved from configuration, and I thought well let's just plug it in there, next to it. There was already an existing logic and structure for initializing setting models, it already does error handling for various cases, so it felt like the right place to add resolving profiles there as well. On top of that, there were already easily-accessible methods for resolving profiles by name (which I believe are not exposed to the "consuming" side).
- I agree that ideally these settings models are "dumb", and that there should be some "application model" that will contain all resolved profiles, but creating models in WinRT is a pain (especially when you're spending 2 days on something that is a typo in the `csproj`...) and I would have no clue where to put it :P So I was being pragmatic there, to not create too many files and classes and whatnot. And in practice there is not _that_ much disparity between the dumb setting model and the "application model"; it's the profile instance (which we'll need anyway for writing back GUIDs for futureproofing), and possibly a list of profile instances for the generated collections like "remaining".

Reply:
That's a completely reasonable explanation. Thank you for indulging me 🙂